Initial functionality

dav3r opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

This PR covers the initial functionality required to generate an Assessor Portal AMI that can be used in [the COOL](https://github.com/cisagov/cool-system).

<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##

This PR is an important part of the work required for https://github.com/cisagov/cool-system/issues/202 and it depends on https://github.com/cisagov/ansible-role-assessor-portal/pull/3.

Note that since the Assessor Portal requires several gigabytes of Docker images to function, those Docker images are pulled in as part of the construction of this AMI (see https://github.com/cisagov/assessor-portal-packer/pull/1/commits/c7a3e80b8240d8a467e85dda7cfaf13a11ef0253, https://github.com/cisagov/assessor-portal-packer/pull/1/commits/90b76b5905fd8224027bf46928352320f5e437ee, and https://github.com/cisagov/assessor-portal-packer/pull/1/commits/a2119926505aaa36fd8023a035036e318debc5da).  This is mainly to avoid having to fetch those Docker images every time a fresh Assessor Portal instance is deployed.

<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such
<!-- as "closes" or "resolves" to auto-close them on merge. -->

## 🧪 Testing ##

This AMI was tested in conjunction with https://github.com/cisagov/cool-assessment-terraform/pull/130, which adds support for these Assessor Portal instances.

Testing confirmed all of the following:
- [x] An Assessor Portal instance could be successfully deployed into the COOL
- [x] The Assessor Portal web application could be successfully started and accessed locally from a browser on the Assessor Portal instance
- [x] Multiple Assessor Portal instances could be successfully deployed in the same COOL assessment environment
- [x] An operator on one Assessor Portal instance could NOT access a running Assessor Portal web application on a different Assessor Portal instance
- [x] The Assessor Portal web application (once started) could NOT be accessed from any other instances in the COOL assessment environment
- [x] The Assessor Portal web application (once started) could NOT be accessed externally via the Internet

<!-- How did you test your changes? How could someone else test this PR? -->

<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

## ✅ Checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

* [x] This PR has an informative and human-readable title.
* [x] Changes are limited to a single goal - _eschew scope creep!_
* [x] All relevant type-of-change labels have been added.
* [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
* [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
* [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
* [x] All new and existing tests pass.

dav3r committed:
Remove example playbook inherited from skeleton

dav3r committed:
Configure build user to have access to all necessary SSM parameters

dav3r committed:
Initialize version to 0.0.1

dav3r committed:
Configure packer to use Debian Bullseye for this image

dav3r committed:
Add necessary requirements and remove unneeded ones from skeleton

Note that backports was needed for Debian Buster, but is not needed here 
since we are using Bullseye.

dav3r committed:
Remove Buster-specific items that are not needed with Bullseye

dav3r committed:
Include the xfce desktop environment, with our COOL customizations

dav3r committed:
Ensure that all cloud-init services are enabled

dav3r committed:
Install VNC so that Guacamole has something to talk to

dav3r committed:
Install the Assessor Portal composition

dav3r committed:
Install useful extra software used by operators

Also add VNC user to docker group for operator convenience

dav3r committed:
Add VNC, Assessory Portal, extras, and cloud-init playbooks

Also, remove Example playbook inherited from skeleton.

dav3r committed:
Update Terraform to find our assessor portal images

dav3r committed:
Replace skeleton info with actual info for this project

dav3r committed:
Bump version from 0.0.1 to 0.0.1-rc.1

dav3r committed:
Add remote state for ansible-role-assessor-portal

dav3r committed:
Attach policies to allow reading assessor portal tarball from S3 buckets

dav3r committed:
Bump version from 0.0.1-rc.1 to 0.0.1-rc.2

dav3r committed:
Specify appropriate BUILD_BUCKET env variable

dav3r committed:
Bump version from 0.0.1-rc.2 to 0.0.1-rc.3

dav3r committed:
Pass build bucket environment variable to playbook.yml

dav3r committed:
Bump version from 0.0.1-rc.3 to 0.0.1-rc.4

dav3r committed:
Properly pass build bucket to assessor_portal ansible role

dav3r committed:
Bump version from 0.0.1-rc.4 to 0.0.1-rc.5

jsf9k commented on src/assessor_portal.yml:
```diff
@@ -0,0 +1,68 @@
+        # The (?!...) jazz is a negative look-ahead assertion.  See
+        # here, for example:
+        # https://docs.python.org/3/library/re.html
+        regexp: '^After=(.*)(?! multi\-user\.target cloud\-final\.service)$'
```
:muscle: regex-fu!

jsf9k commented on src/assessor_portal.yml:
```diff
@@ -0,0 +1,68 @@
+#           path: /home/{{ vnc_username }}/Desktop
+#           state: directory
+#           mode: 0755
+#           owner: "{{ vnc_username }}"
```
You should remove all this commented-out foo if you end up not using it.

dav3r replied:
I did end up removing the systemd bits as I don't think they are necessary for a minimum viable product here.

jsf9k commented on src/packer.json:
```diff
@@ -29,13 +29,13 @@
         "most_recent": true,
         "owners": [
-          "136693071363"
+          "903794441882"
```
Eventually an official Debian 11 AMI will be released, at which point you will probably need to revert this owner account change.  I am now checking monthly that our base AMIs will still match the most recent (there is a recurring calendar item) so this should eventually get picked up by that.

dav3r replied:
Good call - thanks for the reminder.  When I have a few minutes, I will create issues about that for this repo and for [cisagov/pca-gophish-composition-packer](https://github.com/cisagov/pca-gophish-composition-packer), which is in a similar situation.

dav3r replied:
Issues documented in https://github.com/cisagov/assessor-portal-packer/issues/2 and https://github.com/cisagov/pca-gophish-composition-packer/issues/56.

jsf9k reviewed (commented):
A few pre-review comments.  No changes requested.

dav3r committed:
Install desktop shortcut for assessor portal

dav3r committed:
Pull in all Docker images needed for Assessor Portal

dav3r committed:
Bump version from 0.0.1-rc.5 to 0.0.1-rc.6

dav3r committed:
Strip newline from filename

The unintended newline at the end of the filename was preventing this 
shortcut from working as intended.

dav3r committed:
Pull Assessor Portal images in via docker-compose

run.py does not pull in all of the images used by Assessor Portal, so 
docker-compose is used to get the remaining images.

dav3r committed:
Add a comment to state that both image pulls are necessary

dav3r committed:
Bump version from 0.0.1-rc.6 to 0.0.1-rc.7

dav3r committed:
Increase disk size since we are pre-loading in Docker images

8 GB was not large enough since we pull in several GB of Docker images.

dav3r committed:
Bump version from 0.0.1-rc.7 to 0.0.1-rc.8

jsf9k reviewed (approved):
Approved, but please create those issues that you mentioned in response to my earlier reviews.

dav3r commented:
> Approved, but please create those issues that you mentioned in response to my earlier reviews.

Issues documented in https://github.com/cisagov/assessor-portal-packer/issues/2 and https://github.com/cisagov/pca-gophish-composition-packer/issues/56.

mcdonnnj commented on README.md:
```diff
@@ -1,14 +1,9 @@
+software.
 
 ## Pre-requisites ##
 
```
Please update the `Using Your Local Environment` section to include the `BUILD_BUCKET` environment variable as a requirement for the Packer template.

dav3r replied:
Good catch!  Rectified in https://github.com/cisagov/assessor-portal-packer/pull/1/commits/3fad09095574c0b01545002b0eeb24dd69640382.

mcdonnnj reviewed (approved):
Strong work 💪 Please see my comment about a missing documentation update.

dav3r committed:
Mention the BUILD_BUCKET environment variable

Co-Authored-By: Nick M. <mcdonnnj@users.noreply.github.com>
